app/src/main/java/com/sopt/now/presentation/main/mypage/MyPageFragment.kt

```diff
@@ -30,28 +32,27 @@ class MyPageFragment : Fragment() {
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
 
-        memberId = sSharedPreferences.getString(MEMBER_ID, null)
         getUserInfo()
         observeUserInfo()
     }
 
+    private fun getUserInfo() {
+        myPageViewModel.getUserInfo()
+    }
+
     override fun onDestroyView() {
         _binding = null
         super.onDestroyView()
     }
 
-    private fun getUserInfo() {
-        memberId?.let { myPageViewModel.getUserInfo(it) }
-    }
-
     private fun observeUserInfo() {
-        myPageViewModel.status.observe(viewLifecycleOwner) {
-            if (it) {
-                showUserInfo(myPageViewModel.userInfo)
-            } else {
-                toast(myPageViewModel.errorMessage ?: "")
+        myPageViewModel.state.flowWithLifecycle(lifecycle).onEach { state ->
+            when (state) {
+                is UiState.LOADING -> {}
+                is UiState.SUCCESS<*> -> { showUserInfo(state.data as UserInfo) }
```

Comment:
제너릭 타입을 사용하면서 타입 캐스팅이 필요한 것으로 보여요!
해당 타입의 경우 코틀린의 최상위 타입인 Any? 를 반환하기 때문에 타입 캐스팅이 필요 할 겁니다
```suggestion
                is UiState.SUCCESS ->  showUserInfo(state.data)
```
이런식으로도 사용 가능합니다! 첫번째 코드에서는 안전하지 않은 타입 캐스팅 이기 때문에
ClassCastException이 발생할 가능성이 있습니다!

Reply:
![image](https://github.com/NOW-SOPT-ANDROID/jeonghyun-yu/assets/81751105/f704581f-ce76-4a1f-8981-8218a3e35ceb)

저도 제네릭 타입 작성 피하려고 했는데 이렇게 수정해도 컴파일에서 오류가 나네여..

Reply:
이거 마이페이지 뷰모델에서 Uistate의 타입을 지정해주어야해요!

Reply:
우선 util로 선언된 ui state 자체가 제너릭 타입을 사용하여 uiState 에 타입 안정성을 제공해 줍니다! 

```kotlin
sealed class UiState<out T> {
    data object LOADING : UiState<Nothing>()
    data class SUCCESS<T>(
        val data: T? = null,
    ) : UiState<T>()

    data class FAILURE(
        val errorMessage: String,
    ) : UiState<Nothing>()
}
```

다음 MyPageViewModel의 staeflow변수 초기화에 UserInfo타입을 넣어줍니다
```kotlin
    private val _state = MutableStateFlow<UiState<UserInfo>>(UiState.LOADING)
    val state get() = _state.asStateFlow()
```

마지막으로 액티비티에서 flow를 수집해줍니다!
```kotlin
    private fun observeUserInfo() {
        myPageViewModel.state.flowWithLifecycle(lifecycle).onEach { state ->
            when (state) {
                is UiState.LOADING -> {}
                is UiState.SUCCESS -> {
                    state.data?.let { showUserInfo(it) }
                }
                is UiState.FAILURE -> { requireContext().showToast(state.errorMessage) }
            }
        }.launchIn(lifecycleScope)
    }
```

일반적으로 UIstate는 이렇게 사용해요! Uistate의 제너릭 타입을 받아야 타입 안정성이 확보되고 Success시에 어떠한 타입의 데이터를 전달하는지 확실하게 알 수 있어서요!

Reply:
Success시에 데이터를 가져가야하는 마이페이지 뷰모델을 제외하고는 나머진 기존 제너릭타입으로 타입 캐스팅을 해도 무방할것 같긴합니다!

Reply:
아하!! 자세한 설명 감사합니다! 적용해볼게요!!